codex-rs/rmcp-client/src/oauth.rs

```diff
@@ -0,0 +1,823 @@
+use std::fmt;
+use std::fs;
+use std::io::ErrorKind;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use std::time::SystemTime;
+use std::time::UNIX_EPOCH;
+use tracing::warn;
+
+use rmcp::transport::auth::AuthorizationManager;
+use tokio::sync::Mutex;
+
+use crate::find_codex_home::find_codex_home;
+
+const KEYRING_SERVICE: &str = "Codex MCP Credentials";
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct StoredOAuthTokens {
+    pub server_name: String,
+    pub url: String,
+    pub client_id: String,
+    pub token_response: WrappedOAuthTokenResponse,
+}
+
+#[derive(Debug)]
+struct CredentialStoreError(anyhow::Error);
+
+impl CredentialStoreError {
+    fn new(error: impl Into<anyhow::Error>) -> Self {
+        Self(error.into())
+    }
+
+    fn message(&self) -> String {
+        self.0.to_string()
+    }
+
+    fn into_error(self) -> anyhow::Error {
+        self.0
+    }
+}
+
+impl fmt::Display for CredentialStoreError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl std::error::Error for CredentialStoreError {}
+
+trait CredentialStore {
+    fn load(&self, service: &str, account: &str) -> Result<Option<String>, CredentialStoreError>;
+    fn save(&self, service: &str, account: &str, value: &str) -> Result<(), CredentialStoreError>;
+    fn delete(&self, service: &str, account: &str) -> Result<bool, CredentialStoreError>;
+}
+
+struct KeyringCredentialStore;
+
+impl CredentialStore for KeyringCredentialStore {
+    fn load(&self, service: &str, account: &str) -> Result<Option<String>, CredentialStoreError> {
+        let entry = Entry::new(service, account).map_err(CredentialStoreError::new)?;
+        match entry.get_password() {
+            Ok(password) => Ok(Some(password)),
+            Err(keyring::Error::NoEntry) => Ok(None),
+            Err(error) => Err(CredentialStoreError::new(error)),
+        }
+    }
+
+    fn save(&self, service: &str, account: &str, value: &str) -> Result<(), CredentialStoreError> {
+        let entry = Entry::new(service, account).map_err(CredentialStoreError::new)?;
+        entry.set_password(value).map_err(CredentialStoreError::new)
+    }
+
+    fn delete(&self, service: &str, account: &str) -> Result<bool, CredentialStoreError> {
+        let entry = Entry::new(service, account).map_err(CredentialStoreError::new)?;
+        match entry.delete_credential() {
+            Ok(()) => Ok(true),
+            Err(keyring::Error::NoEntry) => Ok(false),
+            Err(error) => Err(CredentialStoreError::new(error)),
+        }
+    }
+}
+
+/// Wrap OAuthTokenResponse to allow for partial equality comparison.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct WrappedOAuthTokenResponse(pub OAuthTokenResponse);
+
+impl PartialEq for WrappedOAuthTokenResponse {
+    fn eq(&self, other: &Self) -> bool {
+        match (serde_json::to_string(self), serde_json::to_string(other)) {
```

Comment:
We can't derive?

Reply:
No because we don't control the original type and it's in another crate. Let me know if there _is_ a way to do that though but I'm not aware of one.